carcara/src/checker/rules/strings.rs

```diff
@@ -323,6 +327,79 @@ fn extract_arguments(t: &Rc<Term>) -> Result<Vec<Rc<Term>>, CheckerError> {
     Ok(args_t.clone())
 }
 
+fn singleton_elim(pool: &mut dyn TermPool, r_list: Vec<Rc<Term>>) -> Rc<Term> {
+    match r_list.len() {
+        1 => r_list[0].clone(),
+        _ => pool.add(Term::Op(Operator::ReConcat, r_list)),
+    }
+}
+
+fn str_fixed_len_re(pool: &mut dyn TermPool, r: Rc<Term>) -> Result<usize, CheckerError> {
+    match r.as_ref() {
+        Term::Op(Operator::ReConcat, args) => {
+            if let [r_1, r_2 @ ..] = &args[..] {
+                let r_2_concat = concat(pool, r_2.to_vec());
+                Ok(str_fixed_len_re(pool, r_1.clone())? + str_fixed_len_re(pool, r_2_concat)?)
+            } else {
+                unreachable!()
+            }
+        }
+        Term::Op(Operator::ReAllChar, _) => Ok(1),
+        Term::Op(Operator::ReRange, _) => Ok(1),
+        Term::Op(Operator::StrToRe, args) => {
+            if let Some(s_1) = args.first() {
+                match s_1.as_ref() {
+                    Term::Const(Constant::String(s)) => Ok(s.len()),
+                    _ => Err(CheckerError::LengthCannotBeEvaluated(r.clone())),
+                }
+            } else {
+                unreachable!()
+            }
+        }
+        Term::Op(Operator::ReUnion, args) => {
+            if let [r_1, r_2 @ ..] = &args[..] {
+                let n = str_fixed_len_re(pool, r_1.clone())?;
+                if r_2.len() == 1 {
+                    match r_2.first().unwrap().as_ref() {
+                        Term::Op(Operator::ReNone, _) => Ok(n),
+                        _ => Err(CheckerError::LengthCannotBeEvaluated(r.clone())),
+                    }
+                } else {
+                    let new_r_2 = pool.add(Term::Op(Operator::ReUnion, r_2.to_vec()));
+                    if str_fixed_len_re(pool, new_r_2)? == n {
+                        Ok(n)
+                    } else {
+                        Err(CheckerError::LengthCannotBeEvaluated(r.clone()))
+                    }
+                }
+            } else {
+                unreachable!()
+            }
+        }
+        Term::Op(Operator::ReIntersection, args) => {
+            if let [r_1, r_2 @ ..] = &args[..] {
+                let n = str_fixed_len_re(pool, r_1.clone())?;
+                if r_2.len() == 1 {
+                    match r_2.first().unwrap().as_ref() {
+                        Term::Op(Operator::ReAll, _) => Ok(n),
+                        _ => Err(CheckerError::LengthCannotBeEvaluated(r.clone())),
+                    }
+                } else {
+                    let new_r_2 = pool.add(Term::Op(Operator::ReIntersection, r_2.to_vec()));
+                    if str_fixed_len_re(pool, new_r_2)? == n {
+                        Ok(n)
+                    } else {
+                        Err(CheckerError::LengthCannotBeEvaluated(r.clone()))
+                    }
+                }
+            } else {
+                unreachable!()
+            }
```

Comment:
Similarly to the `ReConcat` case, in these two cases you are constructing a `re.union`/`re.inter` term with the remaining arguments just to call `str_fixed_re_len` recursively, which is not very efficient (for each argument, you allocate a vector with all the remaining arguments). I think it would be best to call `str_fixed_re_len` on each argument directly to get their lenghts, and then check if they are all the same.

Also, since these two cases are almost identical, differing only in which terms they ignore, it might make sense to put this logic in a separate function and call it in each case.